lib/Runtime/Library/JSON.cpp

```diff
@@ -424,85 +427,114 @@ namespace JSON
+            {
+                PropertyValueInfo info;
+                Js::PropertyString* propertyString = (Js::PropertyString*)key;
+                PropertyValueInfo::SetCacheInfo(&info, propertyString, propertyString->GetLdElemInlineCache(), false);
+                CacheOperators::TryGetProperty<true, false, true, false, true, false, false, true, false>(holder, false, Js::RecyclableObject::FromVar(holder), keyId, &value, scriptContext, nullptr, &info);
+            }
+
+            if (value == nullptr)
+            {
+                if (!Js::JavascriptOperators::GetProperty(Js::RecyclableObject::FromVar(holder), keyId, &value, scriptContext))
+                {
+                    return scriptContext->GetLibrary()->GetUndefined();
+                }
+            }
         }
+        return StrHelper(key, value, holder);
+    }
 
-        if (value == nullptr)
+    inline bool Get_ToJSON(ScriptContext* scriptContext, Js::JavascriptString* key, Js::Var* value, Js::TypeId typeId)
+    {
+        Js::Var toJSON = nullptr;
+        Js::RecyclableObject* object = Js::RecyclableObject::FromVar(*value);
+        while (typeId != Js::TypeIds_Null)
         {
-            if (!Js::JavascriptOperators::GetProperty(Js::RecyclableObject::FromVar(holder), keyId, &value, scriptContext))
+            PropertyQueryFlags result = object->GetPropertyQuery(object, Js::PropertyIds::toJSON, &toJSON, nullptr, scriptContext);
+
+            if (result != PropertyQueryFlags::Property_NotFound && Js::JavascriptConversion::IsCallable(toJSON))
             {
-                return scriptContext->GetLibrary()->GetUndefined();
+                Js::Var values[2];
+                Js::Arguments args(0, values);
+                args.Info.Count = 2;
+                args.Values[0] = *value;
+                args.Values[1] = key;
+
+                Js::RecyclableObject* func = Js::RecyclableObject::FromVar(toJSON);
+                *value = Js::JavascriptFunction::CallFunction<true>(func, func->GetEntryPoint(), args);
+                return true;
             }
+
+            object = object->GetPrototype();
+            typeId = Js::JavascriptOperators::GetTypeId(object);
         }
-        return StrHelper(key, value, holder);
+        return false;
     }
 
     Js::Var StringifySession::StrHelper(Js::JavascriptString* key, Js::Var value, Js::Var holder)
     {
         PROBE_STACK(scriptContext, Js::Constants::MinStackDefault);
         AssertMsg(Js::RecyclableObject::Is(holder), "The holder argument in a JSON::Str function must be an object");
 
-        Js::Var values[3];
-        Js::Arguments args(0, values);
         Js::Var undefined = scriptContext->GetLibrary()->GetUndefined();
+        Js::TypeId id = Js::JavascriptOperators::GetTypeId(value);
 
         //check and apply 'toJSON' filter
         if (Js::JavascriptOperators::IsJsNativeObject(value) || (Js::JavascriptOperators::IsObject(value)))
```

Comment:
Both these functions call GetTypeId- is it worth creating a new function IsNativeOrUserObject which does the following:
Check if Var is recycler aligned- if not, returns false
Check if the type id is either Native object- if so return true
Check if type id > lastPrimitiveType- return true

Not sure if it would make a huge difference but if this path is hot, it might be noticeable

Reply:
@digitalinfinity this is one of the two places we use `Js::JavascriptOperators::IsJsNativeObject` (the one under this function) hence I don't expect perf diff under especially LTO. However, I really liked the idea and will add the interface.